File: benefits-calc-api/BenefitsCalcAPI/Domain/DTOs/EmployeeDto.cs
namespace Domain.DTOs;

public class EmployeeDto
{
    public EmployeeDto(string? firstName, string? lastName, SsnVO ssn)
    {
        FirstName = firstName;
        LastName = lastName;
        Ssn = ssn;
    }

    public EmployeeDto()
    {
        
    }
    public string? FirstName { get; set; }
    public string? LastName { get; set; }
    public string? Ssn { get; set; }
    public int? EmployeeId { get; set; }
}

File: benefits-calc-ui/src/app/benefits/Models/employee-dto.ts
import { IEmployee } from './employee';
export interface IPersonDto {
  firstName: string;
  lastName: string;
  ssn: string;
}
export interface IEmployeeDto extends IPersonDto {
  employeeId?: number;
  dependents: IPersonDto[];
}
export class EmployeeDto implements IEmployeeDto {

  private constructor(
    public firstName: string,
    public lastName: string,
    public ssn: string,
    public dependents: IPersonDto[],
    public employeeId?: number,
  ) {}
  public static fromEntity(employeeEntity: IEmployee): EmployeeDto {
    const {firstName, lastName, ssn, dependents, employeeId} = employeeEntity;
    const dependentsDto: IPersonDto[] = dependents!.map(p => ({
      firstName: p.firstName ?? '',
      lastName: p.lastName ?? '',
      ssn: p.ssn?.fullSsn ?? ''
    })) ?? [];
    return new EmployeeDto(
      firstName!,
      lastName!,
      ssn!.fullSsn,
      dependentsDto,
      employeeId
      );
  }
}
